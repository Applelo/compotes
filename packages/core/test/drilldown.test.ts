import { afterAll, beforeAll, describe, expect, it } from 'vitest'
import type { Browser } from 'playwright'
import { chromium } from 'playwright'
import { destroyComponent } from './helper/destroy'

let browser: Browser

beforeAll(async () => {
  browser = await chromium.launch()
})

afterAll(async () => {
  browser.close()
})

describe('drilldown', async () => {
  it.concurrent('generateId', async () => {
    const page = await browser.newPage()
    await page.goto('http://localhost:3000/drilldown.html')
    const item = page.locator('#c-id-1').first()
    expect(item).toBeDefined()
  })

  it.concurrent('events', async () => {
    const page = await browser.newPage()
    const events = [
      'c.drilldown.update',
      'c.drilldown.next',
      'c.drilldown.back',
      'c.drilldown.reset',
    ]
    const updateEvent = page.waitForEvent('console', msg => msg.text().includes('c.drilldown.update'))
    const nextEvent = page.waitForEvent('console', msg => msg.text().includes('c.drilldown.next'))
    const backEvent = page.waitForEvent('console', msg => msg.text().includes('c.drilldown.back'))
    const resetEvent = page.waitForEvent('console', msg => msg.text().includes('c.drilldown.reset'))
    await page.goto('http://localhost:3000/drilldown.html')

    await page.evaluate(() => {
      const el = document.querySelector('.c-drilldown')
      if (!el)
        return
      [
        'c.drilldown.update',
        'c.drilldown.next',
        'c.drilldown.back',
        'c.drilldown.reset',
      ].forEach((e) => {
        el.addEventListener(e, () => {
          // eslint-disable-next-line no-console
          console.log(e)
        })
      })
    })

    const nextBtn = page.locator('.c-drilldown-next').first()
    const backBtn = page.locator('.c-drilldown-back').first()
    const resetBtn = page.locator('.js-drilldown-reset').first()

    await nextBtn.click()
    await backBtn.click()
    await nextBtn.click()
    await resetBtn.click()

    const result = (await Promise.all([
      updateEvent,
      nextEvent,
      backEvent,
      resetEvent,
    ])).map(item => item.text())

    events.forEach(e => expect(result).toContain(e))
  })

  it.concurrent('destroy', async ({ expect }) => {
    const page = await browser.newPage()
    await page.goto('http://localhost:3000/drilldown.html')
    const { before, after } = await destroyComponent(page)
    expect(before).not.toEqual(after)
    // remove style to prevent issue between local and ci
    expect(after.replace(/ style=".*;"/gm, '')).toMatchSnapshot()
  })
})
