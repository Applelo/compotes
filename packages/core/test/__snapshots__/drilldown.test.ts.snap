// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`drilldown > destroy 1`] = `
"<!DOCTYPE html><html lang="en"><head>
    <link rel="icon" href="data:image/svg+xml,<svg xmlns=%22http://www.w3.org/2000/svg%22 viewBox=%220 0 100 100%22><text y=%22.9em%22 font-size=%2290%22>üçØ</text></svg>">
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Drilldown</title>
    <script type="module" crossorigin="" src="/assets/drilldown-BoIVpoDc.js"></script>
    <link rel="modulepreload" crossorigin="" href="/assets/_parent-PLeNLLls.js">
    <link rel="modulepreload" crossorigin="" href="/assets/index.esm-CjV0_wFH.js">
    <link rel="modulepreload" crossorigin="" href="/assets/accessibility-DQSUv1PD.js">
    <link rel="modulepreload" crossorigin="" href="/assets/animation-CH8HhKvA.js">
    <link rel="stylesheet" crossorigin="" href="/assets/marquee-ybAlwBfl.css">
  </head>
  <body>

    <nav class="c-drilldown">
      <ul class="c-drilldown-menu">
        <li>
          <button class="c-drilldown-next">
            Go to section 1
          </button>
          <ul class="c-drilldown-menu" id="section-1">
            <li>
              <button class="c-drilldown-back">
                Go Back
              </button>
            </li>
            <li>
              <button class="c-drilldown-next">
                Go to section 1 1
              </button>
              <ul class="c-drilldown-menu">
                <li>
                  <button class="c-drilldown-back">
                    Go Back
                  </button>
                </li>
                <li>
                  Item Section 1 1
                </li>
                <li>
                  Item Section 1 1
                </li>
              </ul>
            </li>
            <li>
              Item Section 1
            </li>
          </ul>
        </li>
        <li>
          Hello
        </li>
        <li>
          <button class="c-drilldown-next">
            Go to section 2
          </button>
          <ul class="c-drilldown-menu" id="section-2">
            <li>
              <button class="c-drilldown-back">
                Go Back
              </button>
            </li>
            <li>
              Item Section 2
            </li>
            <li>
              Item Section 2
            </li>
          </ul>
        </li>
        <li>
          <button class="c-drilldown-next">
            Go to section 3
          </button>
          <ul class="c-drilldown-menu" id="section-3">
            <li>
              <button class="c-drilldown-back">
                Go Back
              </button>
            </li>
            <li>
              Item Section 3
            </li>
            <li>
              Item Section 3
            </li>
          </ul>
        </li>
        <li>
          Item
        </li>
      </ul>
    </nav>
    <button class="js-drilldown-reset">Reset</button>
    <button class="js-destroy">
      Destroy
    </button>

  

</body></html>"
`;
